### PR TITLE
Api limit number of alerts

### PR DESCRIPTION
**Pull request by qinxx108**
Add the configuration for introduce an API response limit to protect the service. This limit only has the alerts size limit for now.

### CODE REVIEW

File: api/v1/api.go [hunk 1]
```diff
@@ -305,7 +308,11 @@ func (api *API) listAlerts(w http.ResponseWriter, r *http.Request) {
 	defer alerts.Close()
 
 	api.mtx.RLock()
+	alertCount := 0
 	for a := range alerts.Next() {
+		if api.apiLimit != nil && api.apiLimit.MaxAlertsCount > 0 && api.apiLimit.MaxAlertsCount <= alertCount {
```
Comment (rajagopalanand) [inline api/v1/api.go:313]: Wonder if you can just use the length of res for the check instead of having another new variable

File: api/v2/api.go [hunk 2]
```diff
@@ -266,7 +269,12 @@ func (api *API) getAlertsHandler(params alert_ops.GetAlertsParams) middleware.Re
 	now := time.Now()
 
 	api.mtx.RLock()
+	alertCount := 0
 	for a := range alerts.Next() {
+		if api.apiLimit != nil && api.apiLimit.MaxAlertsCount > 0 && api.apiLimit.MaxAlertsCount <= alertCount {
```
Comment (rajagopalanand) [inline api/v2/api.go:274]: same comment as v1/api.go

Comment (qinxx108) [reply api/v2/api.go]: fixed

File: api/v1/api.go (hunk 1 above)
Comment (qinxx108) [reply api/v1/api.go]: fixed

**Review by emanlodovice — CHANGES_REQUESTED** [review]

File: api/v1/api_test.go [hunk 3]
```diff
@@ -193,81 +193,204 @@ func TestListAlerts(t *testing.T) {
...
+				StartsAt: now.Add(-2 * time.Minute),
+				EndsAt:   now.Add(-time.Minute),
+			},
+		},
+	}
+
+	for i, tc := range []struct {
+		err    bool
+		params map[string]string
+
```
[the remark this answers is not in the document]
Comment (qinxx108) [reply api/v1/api_test.go:379]: done

File: api/v1/api_test.go [hunk 4]
```diff
@@ -193,81 +193,204 @@ func TestListAlerts(t *testing.T) {
...
+			name, ok := a.Labels["alertname"]
+			if ok {
+				anames = append(anames, string(name))
+			}
+		}
+		require.Equal(t, tc.anames, anames, fmt.Sprintf("test case: %d, alert names are not equal", i))
+	}
+}
+
+func TestListAlerts_withAPILimit(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (qinxx108) [reply api/v1/api_test.go:340]: sorry this is mistakenly added. nice catch!

File: api/v2/api.go [hunk 5]
```diff
@@ -267,6 +270,10 @@ func (api *API) getAlertsHandler(params alert_ops.GetAlertsParams) middleware.Re
 
 	api.mtx.RLock()
 	for a := range alerts.Next() {
+		if api.apiLimit != nil && api.apiLimit.MaxAlertsCount > 0 && api.apiLimit.MaxAlertsCount <= len(res) {
```
[the remark this answers is not in the document]
Comment (qinxx108) [reply api/v2/api.go:273]: added

**Review by emanlodovice — APPROVED** [review]
LGTM, thanks @qinxx108

**Pull request closed without merging**